### ISSUE TITLE
Copynet_seq2seq does not correctly call SequenceAccuracy metric

### ISSUE THREAD

**Issue by tomsherborne** [allenai/allennlp#2919]
The implementation of Copynet_seq2seq appears to incorrectly call the SequenceAccuracy metric as a tensor_based_metric. 

SequenceAccuracy expects at least the tensors `predictions (batch_size, k, sequence_length)`, `gold_labels (batch_size, sequence_length)` and copynet_seq2seq passes 
tensors of size `(batch_size, max_predicted_sequence_length)` and `(batch_size, target_sequence_length)`. The Metric expects inputs to be the same size in dimension `-1` and the predictions for all `k` beams. 

**To Reproduce**
Config pasted below. I'm working with GeoQuery data, but this also applies to other .TSV data files.

**Expected behavior**
Model does not crash during validation reporting.

**System (please complete the following information):**
 - OS: Ubuntu 18.04
 - Python version: Python 3.6.3 :: Anaconda custom (64-bit)
 - AllenNLP version: v0.8.3
 - PyTorch version: 1.0.1.post2

**Additional context**
I am happy to work on a fix/PR if this is worth fixing, I'm just not sure where the fix should go. I'm guessing there needs to be a modification to copynet_seq2seq to truncate top_k_predictions to size (batch_size, beam_size, target_sequence_length) and possibly some other adjustments.

Config
```
{
  "dataset_reader": {
    "type": "copynet_seq2seq",
    "source_tokenizer": {
      "type": "word",
      "word_splitter": {
        "type": "just_spaces"
      }
    },
    "target_tokenizer": {
      "type": "word",
      "word_splitter": {
        "type": "just_spaces"
      }
    },
    "source_token_indexers": {
      "tokens": {
        "type": "single_id",
        "namespace": "tokens"
      }
    },
    "target_namespace": "target_tokens"
  },
  "train_data_path": "data/geoquery/funql/geo.en.train",
  "validation_data_path": "data/geoquery/funql/geo.en.val",
  "vocabulary": {
    "directory_path": "data/geoquery/funql/en_vocab_copy"
  },
  "model": {
    "type": "copynet_seq2seq",
    "source_embedder": {
      "type": "basic",
      "token_embedders": {
        "tokens": {
          "type": "embedding",
          "embedding_dim": 150,
          "trainable": true,
          "vocab_namespace": "tokens"
        }
      }
    },
    "encoder": {
      "type": "lstm",
      "input_size": 150,
      "hidden_size": 150,
      "num_layers": 1,
      "bidirectional": false
    },
    "attention": {
      "type": "dot_product"
    },
    "beam_size": 1,
    "max_decoding_steps": 100,
    "target_embedding_dim": 150,
    "copy_token": "@COPY@",
    "source_namespace": "tokens",
    "target_namespace": "target_tokens",
    "tensor_based_metric": {
      "type": "sequence_accuracy"
    },
    "initializer":[
      [
      ".", {"type": "uniform",
            "a": -0.08,
            "b": 0.08}
      ]
    ]
  },
  "iterator": {
    "type": "bucket",
    "padding_noise": 0.0,
    "batch_size": 20,
    "sorting_keys": [
      [
        "target_tokens",
        "num_tokens"
      ]
    ]
  },
  "trainer": {
    "optimizer": {
      "type": "rmsprop",
      "lr": 0.01,
      "alpha": 0.95
    },
    "learning_rate_scheduler": {
      "type": "step",
      "gamma": 0.98,
      "step_size": 5
    },
    "grad_clipping": 5,
    "num_epochs": 250,
    "patience": 10,
    "cuda_device": 0,
    "num_serialized_models_to_keep": 5,
    "should_log_learning_rate": false,
    "should_log_parameter_statistics": false,
    "validation_metric": "+sequence_accuracy"
  },
  "random_seed": 321,
  "numpy_seed": 321,
  "pytorch_seed": 321
}

```

**Comment by kl2806** [comment]
You're right that the sizes do not match up with the `SequenceAccuracy` metric. The default metric used for Copynet is BLEU (https://github.com/allenai/allennlp/blob/master/allennlp/training/metrics/bleu.py#L115) as it is used for text summarization which takes the best prediction and the gold prediction with the sizes documented in the BLEU implementation. This is consistent with how Copynet is calling it.

Can you give some detail on why you want to use it with the `SequenceAccuracy` metric, and how you want to use it with the GeoQuery data?

**Comment by tomsherborne** [comment]
My goal for this is to compare Copy attention to vanilla seq2seq+dot product attention for token based semantic parsing. I replicated the results of [Dong & Lapata 2016](https://arxiv.org/pdf/1601.01280.pdf) using AllenNLP but I wanted to build on this with Copy attention to examine if the Argument anonymisation can be avoided with a copy mechanism. Since BLEU isn't particularly useful for logical forms, I'd prefer to use a sequence accuracy metric to measure if the generated logical form matches the gold logical as a validation metric.

There seems to be a wider inconsistency with the return behaviour of the `get_metric` function in `Metric` class that also breaks `copynet`. BLEU returns `Dict[str, float]`, which Copynet expects but other metrics return only `float`. This breaks [L829](https://github.com/allenai/allennlp/blob/89700de2cab9bb1f71559eae9cd098d88c015dc7/allennlp/models/encoder_decoders/copynet_seq2seq.py#L829) in `CopyNet` which seems easily fixable.

**Comment by kl2806** [comment]
Thanks! I agree that none of the metrics that operate on just the logical forms are perfect, but I don't see how the sequence accuracy would be a better metric than the BLEU score. In code generation, where execution is challenging in practice, BLEU is typically the metric that is used as far as I'm aware (eg. [CONCODE](https://aclweb.org/anthology/D18-1192), [HEARTSTONE](https://arxiv.org/pdf/1603.06744.pdf). I think BLEU or some kind of soft matching like `difflib` would be more informative than sequence accuracy because sequence accuracy assumes you have some alignment which you have to deal with by truncation, etc. as you mentioned if the predicted and gold are different lengths.

If you do want to track the `SequenceAccuracy` on the top k predictions, then I would suggest creating a `self._top_k_predictions_metric` in the `Copynet` model that operates on `top_k_predictions` and also pass in a mask. I'm not sure if adding this metric to `Copynet` and standardizing the metrics would need to go back into the library though since these two metrics aren't meant to be drop in replacements; they expect different sizes / dimensions, and to your point about `get_metrics`, the type annotation is pretty liberal (https://github.com/allenai/allennlp/blob/b90722188a1ff670f5b1cb8a8f2bdb57c42efe1b/allennlp/training/metrics/metric.py#L29). @epwalsh, @brendan-ai2: any thoughts?

**Comment by brendan-ai2** [comment]
@kl2806 , I'm afraid that I don't have anything terribly helpful to add. My understanding is that the variety of types/shapes accepted and returned by the various metrics is why we normally don't make them configurable at the model level. It's usually more robust to have them hardwired in. (It would be nice to improve this part of the library -- something like register metrics and have them called automatically -- but given that one could want metrics on myriad different internals, it's not clear how to actually do this.)

@tomsherborne , the simplest thing for you to do here is probably just to fork the model and call the metric directly. Sorry I can't provide a cleaner solution.

**Comment by tomsherborne** [comment]
Thanks for the help! I'll proceed with my own fork of CopyNet

**Comment by tomsherborne** [comment]
Thanks for the help! I'll proceed with my own fork of CopyNet
